src/main/utility/status_bar.rs

```diff
@@ -0,0 +1,190 @@
+                //   match the previous scroll behaviour).
+                // {save_cursor}
+                //   The cursor is currently at the correct location, so save it for later.
+                // \u{1B}[1;{scroll_height}r
+                //   Set the scroll region to include all rows but the last.
+                // {last_line}{msg}{clear}
+                //   Move to the last row and write the message.
+                // {restore_cursor}
+                //   Restore the cursor position.
+                eprint!(
+                    "{save_cursor}{restore_scroll_region}{restore_cursor}\
+                     {save_cursor}\n{restore_cursor}\
+                     {next_line}{prev_line}\
+                     {save_cursor}\
+                     \u{1B}[1;{scroll_height}r\
+                     {last_line}{msg}{clear}\
+                     {restore_cursor}",
+                    save_cursor = SAVE_CURSOR,
+                    restore_scroll_region = RESTORE_SCROLL_REGION,
+                    restore_cursor = RESTORE_CURSOR,
+                    next_line = NEXT_LINE,
+                    prev_line = PREV_LINE,
+                    scroll_height = rows - 1,
+                    last_line = LAST_LINE,
+                    msg = *state.read().unwrap(),
+                    clear = CLEAR,
+                );
+                let _ = std::io::stderr().flush();
+            }
+            std::thread::sleep(redraw_interval);
+        }
+
+        eprint!(
+            "{save_cursor}{last_line}{clear}{restore_scroll_region}{restore_cursor}",
+            save_cursor = SAVE_CURSOR,
+            last_line = LAST_LINE,
+            clear = CLEAR,
+            restore_scroll_region = RESTORE_SCROLL_REGION,
+            restore_cursor = RESTORE_CURSOR,
+        );
+        let _ = std::io::stderr().flush();
+    }
+
+    /// Stop and remove the status bar.
+    pub fn stop(&mut self) {
+        self.stop_flag
+            .swap(true, std::sync::atomic::Ordering::Relaxed);
+        if let Some(handle) = self.thread.take() {
+            if let Err(e) = handle.join() {
+                log::warn!("Progress bar thread did not exit cleanly: {:?}", e);
+            }
+        }
+    }
+
+    /// Update the state of the status bar.
+    pub fn mutate_state(&self, f: impl FnOnce(&mut T)) {
+        f(&mut *self.state.write().unwrap())
+    }
+}
+
+pub struct ShadowStatusBarState {
+    start: std::time::Instant,
+    current: SimulationTime,
+    end: SimulationTime,
+}
+
+impl std::fmt::Display for ShadowStatusBarState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let frac = self.current.as_millis() as f32 / self.end.as_millis() as f32;
+        let sim_current = self.current.as_secs();
+        let sim_end = self.end.as_secs();
+        write!(
+            f,
+            "{}% — simulated: {}/{}s, realtime: {}s",
+            (frac * 100.0).round() as i8,
+            sim_current,
```

Comment:
Optional (could be a later improvement), but it might be nice to display in HH:MM:SS format for longer simulations. `shadow_logger.rs` has a `TimeParts` helper that might be helpful here.

Reply:
Added.